src/internal/ppsutil/decoder.go

```diff
@@ -127,3 +131,182 @@ func (r *PipelineManifestReader) NextCreatePipelineRequest() (*ppsclient.CreateP
+// PipelineManifestReader.
+type SpecReader struct {
+	decoder *yaml.Decoder
+	next    func() (string, error)
+}
+
+// NewSpecReader returns a SpecReader which reads from r.
+func NewSpecReader(r io.Reader) *SpecReader {
+	return &SpecReader{decoder: yaml.NewDecoder(r)}
+}
+
+// Next returns the next pipeline spec as a JSON string.  If there are multiple
+// YAML documents, it will return them one at a time.  If there is an array of
+// specs, it will return them one at a time.
+func (r *SpecReader) Next() (string, error) {
+	if r.next != nil {
+		result, err := r.next()
+		switch {
+		case errors.Is(err, io.EOF):
+			return "", err
+		case err != nil:
+			return "", errors.Wrapf(err, "malformed pipeline spec")
+		default:
+			return result, nil
+		}
+	}
+	// no list is in progress: parse the next document as either a spec or a
+	// list of specs.
+	var holder yaml.Node
+	if err := r.decoder.Decode(&holder); err != nil {
+		if errors.Is(err, io.EOF) {
+			return "", io.EOF
+		}
+		return "", errors.Wrapf(err, "malformed pipeline spec")
+	}
+	if holder.Kind != yaml.DocumentNode {
+		return "", errors.Errorf("unexpected YAML kind %v", holder.Kind)
+	}
+	content := holder.Content
+	if len(content) != 1 {
+		return "", errors.Errorf("expected a single YAML document; got %d", len(content))
+	}
+	switch content[0].Kind {
+	case yaml.SequenceNode:
+		index := 0
+		r.next = func() (string, error) {
+			index++
+			if index >= len(content[0].Content) {
+				r.next = nil
+			}
+			return r.convertRequest(content[0].Content[index-1])
+		}
+		return r.Next()
+	case yaml.MappingNode:
+		return r.convertRequest(holder.Content[0])
+	default:
+		return "", errors.Errorf("expected a mapping or a sequence; got %v", content[0].Kind)
+	}
+}
+
+func (r *SpecReader) convertRequest(n *yaml.Node) (string, error) {
+	object, err := yamlToJSON(n)
+	if err != nil {
+		return "", errors.Wrap(err, "could not convert YAML to JSON object")
+	}
+	return r.validateRequest(object)
+}
+
+func (r *SpecReader) validateRequest(jsObj any) (string, error) {
+	b, err := json.Marshal(jsObj)
+	if err != nil {
+		return "", errors.Wrapf(err, "could not marshal %v as JSON", jsObj)
+	}
+	var req pps.CreatePipelineRequest
+	if err := protojson.Unmarshal(b, &req); err != nil {
+		return "", errors.Wrapf(err, "could not unmarshal %s as CreatePipelineRequest", string(b))
+	}
+
+	if req.Pipeline == nil {
+		return "", errors.New("no `pipeline` specified")
+	}
+	if req.Pipeline.Name == "" {
+		return "", errors.New("no pipeline `name` specified")
+	}
+
+	if req.Transform != nil && req.Transform.Image != "" {
+		if !strings.Contains(req.Transform.Image, ":") {
+			fmt.Fprintf(os.Stderr,
+				"WARNING: please specify a tag for the docker image in your transform.image spec.\n"+
+					"For example, change 'python' to 'python:3' or 'bash' to 'bash:5'. This improves\n"+
+					"reproducibility of your pipelines.\n\n")
+		} else if strings.HasSuffix(req.Transform.Image, ":latest") {
```

Comment:
It feels kind of weird to me that there's one particular tag that we attach semantics and a warning to. do we already warn about this? Just my two cents

Reply:
Yes, this was an existing validity check in `PipelineManifestReader`.